File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    static boolean[] v = new boolean[8];
    static int ans = 0;
    public int solution(int k, int[][] dungeons) {
        DFS(0, k, dungeons);
        
        return ans;
    }
    
    static void DFS(int count, int energy, int[][] dungeons){
        ans = Math.max(count , ans);
        if(energy == 0 || count == dungeons.length){
            return;
        }
        
        for(int i = 0; i < dungeons.length; i++){
            if(v[i] || energy < dungeons[i][0]) continue;
            v[i] = true;
            DFS(count + 1, energy - dungeons[i][1], dungeons);
            v[i] = false;
        }
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
ans, v = 0, []
def solution(k, dungeons):
    global v
    v = [False] * len(dungeons)
    find(k, dungeons, 0)
    return ans


def find(fatigue, dungeons, steps):
    global v, ans
    ans = max(ans, steps)
    for i in range(len(dungeons)):
        if v[i]:
            continue
        required, cost = dungeons[i]
        
        if fatigue < required:
            continue
            
        v[i] = True
        find(fatigue - cost, dungeons, steps + 1)
        v[i] = False